torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp

```diff
@@ -0,0 +1,296 @@
+  for (auto concrete_id : dom_set) {
+    // If this concrete domain has a constant extent, check if it
+    // matches with the known constant extent.
+    auto it = constant_extent_map_.find(concrete_id);
+    if (it != constant_extent_map_.end()) {
+      const auto& const_extent_set = it->second;
+      // If multiple constants are detected, it's not exact.
+      if (const_extent_set.size() > 1) {
+        all_equal = false;
+        break;
+      }
+      auto this_const = *(const_extent_set.begin());
+      // known_const is initialized to -1
+      if (known_const == -1) {
+        known_const = this_const;
+      } else if (known_const == this_const) {
+        // Matched with previously known const. The extent of this
+        // domain must be equal to that's previously known.
+        continue;
+      } else {
+        // Unmatched. This dom_set extents may not be unique.
+        all_equal = false;
+        break;
+      }
+    }
+
+    // At this point, it still remains undetermined whether this id
+    // matches with those previously looked at. Constant check failed,
+    // but symbolic matching may succeed.
+    if (!equalDim(
+            known_dimension, gpu_lower->lowerValue(concrete_id->extent()))) {
+      all_equal = false;
+      break;
+    }
+  }
+
+  // If all_equal is still true, the dimension of this paralel type
+  // must be exact.
+  if (all_equal) {
+    exact_types_.insert(pt);
+  }
+  // Use the const value, if found, as its dimension
+  if (all_equal && known_const != -1) {
+    dim_map_.insert({pt, ir_builder.create<kir::Int>(known_const)});
+  } else {
+    dim_map_.insert({pt, kir::NamedScalar::getParallelDim(pt)});
+  }
+}
+
+kir::Val* ParallelDimensionMap::get(ParallelType pt) const {
+  TORCH_INTERNAL_ASSERT(isParallelTypeThread(pt), "Invalid ParallelType: ", pt);
+  auto it = dim_map_.find(pt);
+  if (it == dim_map_.end()) {
+    return nullptr;
+  } else {
+    return it->second;
+  }
+}
+
+bool ParallelDimensionMap::isExact(ParallelType pt) const {
+  return exact_types_.find(pt) != exact_types_.end();
+}
+
+IterDomain* ParallelDimensionMap::getCAMappedConcreteDomain(IterDomain* id) {
+  const auto gpu_lower = GpuLower::current();
+  const auto& ca_map = gpu_lower->caIndexMap();
+  return ca_map.getConcreteMappedID(id);
+}
+
+// Symbolically compares equality of two KIR vals. Comparison is done
+// conservatively, so returning false does not guarantee non-equality.
+bool ParallelDimensionMap::equalDim(kir::Val* dim1, kir::Val* dim2) {
+  TORCH_INTERNAL_ASSERT(dim1 != nullptr && dim2 != nullptr);
+
+  if (dim1 == dim2) {
+    return true;
+  }
+
+  // When Both are Int, they are same if both have the same constant
+  auto dim1_int = dynamic_cast<kir::Int*>(dim1);
+  auto dim2_int = dynamic_cast<kir::Int*>(dim2);
+  if (dim1_int && dim2_int) {
+    if (dim1_int->isConst() && dim2_int->isConst()) {
+      return dim1_int->value() == dim2_int->value();
```

Comment:
nitpick: may need an expression evaluator for constant expressions, as `constCheck` will be recursive.

Reply:
`isConst` here doesn't use `constCheck`.

Reply:
Oh I see. Thanks.